src/orion/core/worker/primary_algo.py

```diff
@@ -79,64 +119,133 @@ def suggest(self, num):
+            # Register the equivalence between these trials.
+            self.registry_mapping.register(original, transformed_trial)
         return trials
 
-    def observe(self, trials):
+    def observe(self, trials: list[Trial]) -> None:
         """Observe evaluated trials.
 
         .. seealso:: `orion.algo.base.BaseAlgorithm.observe`
         """
-        transformed_trials = []
+        # For each trial in the original space, find the suggestions from the algo that match it.
+        # Then, we make the wrapped algo observe each equivalent trial, with the updated status.
         for trial in trials:
-            self._verify_trial(trial, space=self.space)
-            transformed_trials.append(self.transformed_space.transform(trial))
-        self.algorithm.observe(transformed_trials)
+            # Update the status of this trial in the registry (if it was suggested), otherwise set
+            # it in the registry (for example in testing when we force the algo to observe a trial).
+            self.registry.register(trial)
+
+            # Get the known transformed trials that correspond to this original.
+            transformed_trials = self.registry_mapping.get_trials(trial)
 
-    def has_suggested(self, trial):
+            # Also transfer the status and results of `trial` to the equivalent transformed trials.
+            transformed_trials = [
+                _copy_status_and_results(
+                    original_trial=trial, transformed_trial=transformed_trial
+                )
+                for transformed_trial in transformed_trials
+            ]
+
+            if not transformed_trials:
+                # `trial` is a new, original trial that wasn't suggested by the algorithm. (This
+                # might happen when an insertion is done according to @bouthilx)
+                transformed_trial = self.transformed_space.transform(trial)
+                transformed_trial = _copy_status_and_results(
+                    original_trial=trial, transformed_trial=transformed_trial
+                )
+                transformed_trials = [transformed_trial]
+                logger.debug(
+                    f"Observing trial {trial} (transformed as {transformed_trial}), even "
+                    f"though it wasn't suggested by the algorithm."
+                )
+                # NOTE: @lebrice Here we don't want to store the transformed trial in the
+                # algo's registry (by either calling `self.algorithm.register(transformed_trial)` or
+                # `self.algorithm.registry.register(transformed_trial))`, because some algos can't
+                # observe trials that they haven't suggested. We'd also need to perform all the
+                # logic that the algo did in `suggest` (e.g. store it in a bracket for HyperBand).
+                # Therefore we only register it in the wrapper, and store the equivalence between
+                # these two trials in the registry mapping.
```

Comment:
I think this is the correct way to handle this situation. And to be coherent, this suggests that we should not register the trials during `suggest` in the algorithm registry. It should be the algorithm's responsibility.